enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/BoltInteraction.java

```diff
@@ -0,0 +1,318 @@
+    {
+        return subject.username;
+    }
+
+    @Override
+    public void tearDown() throws Throwable
+    {
+        for ( BoltSubject subject : subjects.values() )
+        {
+            subject.client.disconnect();
+        }
+        subjects.clear();
+        server.graphDatabaseService().shutdown();
+        Path IMPERMANENT_DB_ROLES_PATH = Paths.get( "target/test-data/impermanent-db/data/dbms/roles" );
+        if ( Files.exists( IMPERMANENT_DB_ROLES_PATH ) )
+        {
+            Files.delete( IMPERMANENT_DB_ROLES_PATH );
+        }
+    }
+
+    @Override
+    public void assertAuthenticated( BoltSubject subject )
+    {
+        assertTrue( "Should be authenticated", subject.isAuthenticated() );
+    }
+
+    @Override
+    public void assertPasswordChangeRequired( BoltSubject subject )
+    {
+        assertTrue( "Should need to change password", subject.passwordChangeRequired() );
+    }
+
+    @Override
+    public void assertUnauthenticated( BoltSubject subject )
+    {
+        assertFalse( "Should not be authenticated", subject.isAuthenticated() );
+    }
+
+    private static BoltResult collectResults( Connection client ) throws Exception
+    {
+        Message message = TransportTestUtil.recvOneMessage( client );
+        List<String> fieldNames = new ArrayList<>();
+        List<Map<String,Object>> result = new ArrayList<>();
+
+        if ( message instanceof SuccessMessage )
+        {
+            Map<String,Object> metadata = ((SuccessMessage) message).meta();
+            fieldNames = (List<String>) metadata.get( "fields" );
+        }
+        else if ( message instanceof FailureMessage )
+        {
+            FailureMessage failMessage = ((FailureMessage) message);
+            client.send( TransportTestUtil.chunk( reset() ) );
+            collectResults( client );
+            throw new AuthenticationException( failMessage.status(), failMessage.message() );
+        }
+
+        do
+        {
+            message = TransportTestUtil.recvOneMessage( client );
+            if ( message instanceof RecordMessage )
+            {
+                Object[] row = ((RecordMessage) message).record().fields();
+                Map<String,Object> rowMap = new HashMap<>();
+                for ( int i = 0; i < row.length; i++ )
+                {
+                    rowMap.put( fieldNames.get( i ), row[i] );
+                }
+                result.add( rowMap );
+            }
+        }
+        while ( !(message instanceof SuccessMessage) && !(message instanceof FailureMessage) );
+
+        if ( message instanceof FailureMessage )
+        {
+            FailureMessage failMessage = ((FailureMessage) message);
+            client.send( TransportTestUtil.chunk( reset() ) );
+            TransportTestUtil.recvOneMessage( client );
+            throw new AuthenticationException( failMessage.status(), failMessage.message() );
+        }
+
+        return new BoltResult( result );
+    }
```

Comment:
Although I see particular reason to chase performance in this test code, I just want to mention that the transformation between the streams could have been done lazily =)